packages/govuk-frontend/src/govuk/components/character-count/template.njk

```diff
@@ -2,85 +2,100 @@
 {% from "../textarea/macro.njk" import govukTextarea %}
 {% from "../hint/macro.njk" import govukHint %}
 
+{#-
+  If the limit is set in JavaScript, we won't be able to interpolate the message
+  until JavaScript, so we only set a text if the `maxlength` or `maxwords` options
+  were provided to the macro.
+#}
 {%- set hasNoLimit = (not params.maxwords and not params.maxlength) -%}
+{%- set textareaDescriptionLength = params.maxwords or params.maxlength -%}
+{%- set textareaDescriptionText = params.textareaDescriptionText or 'You can enter up to %{count} ' + ('words' if params.maxwords else 'characters') -%}
+{%- set textareaDescriptionTextNoLimit = textareaDescriptionText | replace('%{count}', textareaDescriptionLength) if not hasNoLimit -%}
 
-<div class="govuk-character-count" data-module="govuk-character-count"
-  {%- if params.maxlength %} data-maxlength="{{ params.maxlength }}"{% endif %}
-  {%- if params.threshold %} data-threshold="{{ params.threshold }}"{% endif %}
-  {%- if params.maxwords %} data-maxwords="{{ params.maxwords }}"{% endif %}
-  {#-
-    Without maxlength or maxwords, we can't guess if the component will count words or characters.
-    We can't guess a default textarea description to be interpolated in JavaScript
-    once the maximum gets configured there.
-    So we only add the attribute if a textarea description was explicitely provided.
-  #}
-  {%- if hasNoLimit and params.textareaDescriptionText %}
-    {{- govukI18nAttributes({
```

Comment:
By looping params and appending to `attributes` we can't use `govukI18nAttributes()` anymore

But since Nunjucks can both push to and loop arrays (as if they were objects) using `Object.entries()` format, it could be a great way to merge default `attributes` using https://github.com/alphagov/govuk-frontend/issues/2917

This comes from a little known Nunjucks [**for** loop](https://mozilla.github.io/nunjucks/templating.html#for) feature where these are interchangeable:

## Attributes as plain object

```njk
{{ govukButton({
  text: "Press me",
  attributes: {
    data-example: "value"
  }
}) }}
```

## Attributes as `Object.entries()` array

```njk
{{ govukButton({
  text: "Press me",
  attributes: [
    ["data-example": "value"]
  ]
}) }}
```

Reply:
There's some really clever stuff going on here – not only treating 'arrays of arrays' as objects but also the later re-assignment of `attributes` and the (ab)use of [the comma operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comma_operator), which was new to me!

However, I think the features we're (ab)using are just a little too obscure. I think it would be hard for others who aren't familiar with them to understand what's going on, which makes it hard to read and to make changes.

It did however get me thinking about an alternative approach – what if we allowed for `formGroup.attributes` to be either an object _or a string_ in the textarea component? Removing the requirement for us to pass an object allows the character count to effectively 'take ownership' of the attributes on the form group. It also means we can continue to use the `govukI18nAttributes` macro.

I've pushed a branch to try and demonstrate what this might look like – take a look at https://github.com/alphagov/govuk-frontend/compare/character-count-form-group...spike-pass-attributes-as-string?expand=1 and see what you think?

Reply:
Ha, that's fair. Every Nunjucks service I've ever worked on (ab)uses the comma operator 😆

Will take a look 👀

Reply:
Yep that's a nice enhancement that gets around my "appending to objects" blocker 🙌

Got my support, how do you want to proceed?

Reply:
I'm not sure – I think we will want to eventually make this consistent, so anywhere you can pass `attributes` as an object you can also pass a string? I don't think we have to do that to unblock this PR, but we should probably raise an issue and see where else we can apply the same 'fix'.

I'm also wondering how it should be documented given we only allow for a single 'type' in the YAML – possibly we keep it as a `type: object` but mention you can pass a string in the description?

Happy for you to pick this up and run with it if you're happy to update this PR to adopt that approach?

Reply:
I've put the associated issues back into in progress for now.

Reply:
Issue raised for this in https://github.com/alphagov/govuk-frontend/issues/4666

Reply:
This was a great suggestion by the way 🙌

Made possible by https://github.com/alphagov/govuk-frontend/pull/4742 and added in https://github.com/alphagov/govuk-frontend/pull/4566/commits/779004c4fa1f11a2af517e79fea55fce565f442d